### PR TITLE
fix(api): Error instead of infinite-looping when disposal_volume is too high to make progress

### LINKED ISSUE #6170: bug: Inappropriate disposal or air_gap volume causes hang in transfers.py
# Overview

Accidentally including an inappropriate disposal volume (disposal_volume=pipette_volume) causes a hang in transfers. This leads to infinite simulation times and difficulty troubleshooting. 

It appears to be due to bad assumptions in this `while` statement in `transfers.py`:

https://github.com/Opentrons/opentrons/blob/ee1b5041e69dc930a91ed330fbef17d84a0ae009/api/src/opentrons/protocol_api/transfers.py#L564-L569

I'm not familiar with the code base, but it seems like you want to handle the case where the summed value exceeds `self._max_volume`.

# Steps to reproduce

Minimal example protocol:

```python
from opentrons import protocol_api

metadata = {'apiLevel': '2.5',
            'author': 'Jon Sanders'}

def run(protocol: protocol_api.ProtocolContext):

    protocol.home()

    tiprack_reagents = protocol.load_labware('opentrons_96_tiprack_20ul', 
                                             11)
    reagents = protocol.load_labware('opentrons_96_aluminumblock_generic_pcr_strip_200ul',
                                     1, 'reagents')
    samples = protocol.load_labware('biorad_96_wellplate_200ul_pcr',
                                   4, 'samples')

    pipette_right = protocol.load_instrument('p10_multi', 
                                            'right',
                                            tip_racks=[tiprack_reagents])


    pipette_right.distribute(10,
                         reagents['A1'],
                         samples.wells(),
                         disposal_volume=10,
                         new_tip='once')

```

Out from kernel interrupt:

```
---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
<ipython-input-8-b50bc670336c> in <module>
      1 protocol_file = open('../test.py', 'r')
      2 
----> 3 runlog, _bundle = simulate(protocol_file)
      4 
      5 protocol_file.close()

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/simulate.py in simulate(protocol_file, file_name, custom_labware_paths, custom_data_paths, propagate_logs, hardware_simulator_file_path, log_level)
    342         scraper = CommandScraper(stack_logger, log_level, context.broker)
    343         try:
--> 344             execute.run_protocol(protocol, context)
    345             if isinstance(protocol, PythonProtocol)\
    346                and protocol.api_level >= APIVersion(2, 0)\

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/execute.py in run_protocol(protocol, context)
    109     if isinstance(protocol, PythonProtocol):
    110         if protocol.api_level >= APIVersion(2, 0):
--> 111             _run_python(protocol, context)
    112         else:
    113             raise RuntimeError(

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/execute.py in _run_python(proto, context)
     86     new_globs['__context'] = context
     87     try:
---> 88         exec('run(__context)', new_globs)
     89     except (SmoothieAlarm, asyncio.CancelledError, ExecutionCancelledError):
     90         # this is a protocol cancel and shouldn't have special logging

<string> in <module>

<protocol> in run(protocol)

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/util.py in decorated(*args, **kwargs)
    612                 do_publish(
    613                     broker, command, f, 'before', None, meta, *args, **kwargs)
--> 614             res = f(*args, **kwargs)
    615             if after:
    616                 do_publish(

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/util.py in _check_version_wrapper(*args, **kwargs)
    368                     f'must increase your API version to {added_in} to '
    369                     'use this functionality.')
--> 370             return decorated_obj(*args, **kwargs)
    371 
    372         return _check_version_wrapper

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/instrument_context.py in distribute(self, volume, source, dest, *args, **kwargs)
    848             'disposal_volume', self.min_volume)
    849         kwargs['mix_after'] = (0, 0)
--> 850         return self.transfer(volume, source, dest, **kwargs)
    851 
    852     @cmds.publish.both(command=cmds.consolidate)

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/util.py in decorated(*args, **kwargs)
    612                 do_publish(
    613                     broker, command, f, 'before', None, meta, *args, **kwargs)
--> 614             res = f(*args, **kwargs)
    615             if after:
    616                 do_publish(

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/util.py in _check_version_wrapper(*args, **kwargs)
    368                     f'must increase your API version to {added_in} to '
    369                     'use this functionality.')
--> 370             return decorated_obj(*args, **kwargs)
    371 
    372         return _check_version_wrapper

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/instrument_context.py in transfer(self, volume, source, dest, trash, **kwargs)
   1025                                       self.api_version, kwargs['mode'],
   1026                                       transfer_options)
-> 1027         self._execute_transfer(plan)
   1028         return self
   1029 

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/instrument_context.py in _execute_transfer(self, plan)
   1029 
   1030     def _execute_transfer(self, plan: transfers.TransferPlan):
-> 1031         for cmd in plan:
   1032             getattr(self, cmd['method'])(*cmd['args'], **cmd['kwargs'])
   1033 

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/transfers.py in __iter__(self)
    426         yield from {TransferMode.CONSOLIDATE: self._plan_consolidate,
    427                     TransferMode.DISTRIBUTE: self._plan_distribute,
--> 428                     TransferMode.TRANSFER: self._plan_transfer}[self._mode]()
    429         if self._strategy.new_tip == types.TransferTipPolicy.ONCE:
    430             if self._strategy.drop_tip_strategy == DropTipStrategy.RETURN:

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/transfers.py in _plan_distribute(self)
    564                 while (sum(a[0] for a in asp_grouped) +
    565                        self._strategy.disposal_volume +
--> 566                        self._strategy.air_gap +
    567                        current_xfer[0]) <= self._max_volume:
    568                     asp_grouped.append(current_xfer)

~/miniconda3/envs/py3/lib/python3.7/site-packages/opentrons/protocol_api/transfers.py in <genexpr>(.0)
    562             asp_grouped: List[Tuple[float, Well]] = []
    563             try:
--> 564                 while (sum(a[0] for a in asp_grouped) +
    565                        self._strategy.disposal_volume +
    566                        self._strategy.air_gap +

KeyboardInterrupt:
```

# Current behavior

Simulation runs indefinitely.

# Expected behavior

Simulation errors.

### PR DESCRIPTION
**Pull request by SyntaxColoring**
# Overview

This PR turns code like this into an error:

```python
# disposal_volume too high.
# There would be no room left in the tip to move liquid to dests.
p300.distribute(123, source, dests, disposal_volume=300)
```

The protocol API's current behavior for code like this depends on the exact circumstances. For example, sometimes, it hangs in an infinite loop (as reported in #6170). Other times, it seems to treat the `distribute()` step as a no-op.

This PR turns one specific variant into an error, and moves towards characterizing other variants.

# Changelog

* Fix #6170. Before, the test protocol in that ticket would infinite-loop. Now, it raises a `ValueError` internally.

* Leave these related bugs *unfixed for now*:

  * When the pipette tip has a lower maximum volume than the pipette itself, it seems like an invalid disposal volume can make the `distribute()` effectively *no-op*. It picks up a tip and then drops it, without any sub-steps.

    This PR's unit tests caught this by chance. They're `xfail`'d.

    If we merge this PR without a fix for this bug, I'll ticket this bug separately.

  * Even with this PR, another variant of #6170, calling `transfer()` instead of `distribute()`, still hangs.

     ```python
     metadata = {"apiLevel": "2.7"}

     def run(protocol):
         labware = protocol.load_labware('nest_12_reservoir_15ml', 1)
         tip_rack = protocol.load_labware('opentrons_96_filtertiprack_200ul', 2)
         pipette = protocol.load_instrument('p300_single', mount='left', tip_racks=[tip_rack])
         pipette.transfer(10, labware.wells()[0], labware.wells()[1], disposal_volume=200)
     ```

     I haven't looked into why.

     I found this by manual fiddling, so it's not currently represented in the unit tests.

     If we merge this PR without a fix for this bug, I'll ticket this bug separately.

# Risk assessment

Medium.

There's a risk that this PR accidentally changes edge-case behavior in runnable protocols without an `apiLevel` bump.

This PR is intended to maintain the exact current behavior of all runnable protocols. It should merely improve how the error is presented for certain non-runnable protocols. (Instead of infinite-looping, there should be an error message.)

But suppose someone's written a "wrong," but runnable, protocol. The protocol provides an invalid value to one of our API functions, but the function currently happens not to raise any error. If this PR accidentally *does* make it raise an error, that would break our API versioning policy.

It's possible that this is the case, because:

* I'm not confident I fully understand how `_expand_for_volume_constraints()` is called.
* Although the test that this PR introduces is a good start, the current edge-case behaviors of `transfer()`, `distribute()`, and `consolidate()` don't seem well-characterized.
  * Evidently, there are variants of this bug other than exactly what's reported in #6170. This test catches some variants, but not all.
  * I didn't have a great way of telling Pytest to expect a test to time out. I'm *assuming* that, without the fix, *all* the non-`xfail`'d test parametrizations would time out *if they had a chance to run.* If this assumption is wrong, this PR could be changing the behavior of runnable protocols.

# Review requests

Given the deferred work that I mentioned in "Changelog," and the risks I mentioned in "Risk assessment," are we comfortable merging this as-is? Or should I try to understand this code more before I poke it with a stick?

### CODE REVIEW

File: api/tests/opentrons/protocols/advanced_control/test_transfers.py [hunk 1]
```diff
@@ -1202,114 +1202,74 @@ def test_zero_volume_causes_transfer_of_disposal_vol(_instr_labware):
         assert step == expected
 
 
-def test_blowout_to_source(_instr_labware):
```
Comment (SyntaxColoring) [inline api/tests/opentrons/protocols/advanced_control/test_transfers.py]: This looks like a rebase mistake, of some kind.

**Comment by SyntaxColoring** [comment]
https://github.com/Opentrons/opentrons/issues/7243 looks like it could be related.

File: api/tests/opentrons/protocols/advanced_control/test_transfers.py [hunk 2]
```diff
@@ -1321,3 +1321,77 @@ def test_blowout_to_dest(_instr_labware):
...
+        # not just distributes.
+        plan = tx.TransferPlan(
+            volume=1.2345,
+            sources=labware.rows()[0][0],
+            dests=labware.rows()[0],
+            instr=pipette,
+            max_volume=tip_max_volume,
+            api_version=context.api_version,
+            options=options)
+        for step in plan:
```
Comment (amitlissack) [inline api/tests/opentrons/protocols/advanced_control/test_transfers.py]: Can we do `list(plan)` instead of the loop?

File: api/tests/opentrons/protocols/advanced_control/test_transfers.py [hunk 3]
```diff
@@ -1321,3 +1321,77 @@ def test_blowout_to_dest(_instr_labware):
...
+
+@pytest.mark.parametrize(
+    'pipette_name,tip_rack_name,tip_max_volume,too_high_disposal_volume',
+    [
+        # disposal_volume == pipette max == tip max
+        ('p20_single_gen2', 'opentrons_96_filtertiprack_20ul', 20, 20),
+        ('p300_single', 'opentrons_96_tiprack_300ul', 300, 300),
+
+        # pipette max != tip max, disposal_volume == tip max
+        # todo(mm, 2021-03-10): These fail unexpectedly, apparently a bug.
```
Comment (amitlissack) [inline api/tests/opentrons/protocols/advanced_control/test_transfers.py]: Where do these skipped tests fail?

File: api/tests/opentrons/protocols/advanced_control/test_transfers.py [hunk 4]
```diff
@@ -1321,3 +1321,77 @@ def test_blowout_to_dest(_instr_labware):
...
+    # actually transfer liquid and make progress.
+    #
+    # (As opposed to getting stuck in an infinite loop, or returning an empty
+    # list of steps.)
+
+    # todo(mm, 2021-03-09): Also test that TransferPlan raises when air_gap,
+    # or air_gap + disposal_volume, is too high.
+
+    # Boilerplate: TransferPlan wants an InstrumentContext.
+    context = papi.ProtocolContext(
```
Comment (amitlissack) [inline api/tests/opentrons/protocols/advanced_control/test_transfers.py]: You can skip this boilerplate by using the `ctx` fixture defined in `conftest.py`.

File: api/tests/opentrons/protocols/advanced_control/test_transfers.py (hunk 3 above)
Comment (SyntaxColoring) [reply api/tests/opentrons/protocols/advanced_control/test_transfers.py]: These tests fail because the function under test returns without raising the expected `ValueError`.

Here's what it returns, as printed by this test's `print()` statement:

```
--- Captured stdout call ---
{'method': 'pick_up_tip', 'args': [], 'kwargs': {}}
{'method': 'drop_tip', 'args': [], 'kwargs': {}}
```

File: api/src/opentrons/protocols/advanced_control/transfers.py [hunk 5]
```diff
@@ -600,6 +605,11 @@ def _expand_for_volume_constraints(
         """ Split a sequence of proposed transfers if necessary to keep each
         transfer under the given max volume.
         """
+
+        if max_volume <= 0:
+            raise ValueError(
```
Comment (mcous) [inline api/src/opentrons/protocols/advanced_control/transfers.py:610]: Should we take this opportunity to define a more specific error type for transfer build errors?

File: api/src/opentrons/protocols/advanced_control/transfers.py [hunk 6]
```diff
@@ -600,6 +605,11 @@ def _expand_for_volume_constraints(
         """ Split a sequence of proposed transfers if necessary to keep each
         transfer under the given max volume.
         """
+
+        if max_volume <= 0:
```
Comment (mcous) [inline api/src/opentrons/protocols/advanced_control/transfers.py:609]: I think this might make our future rework a little bit harder, but it's probably worth guarding this in an `apiVersion` check, right?

Comment (SyntaxColoring) [reply api/src/opentrons/protocols/advanced_control/transfers.py:609]: The original hope was that this PR would be narrowly scoped to just the variants of this bug whose fixes wouldn't need an `apiVersion` check. See "Risk assessment" in my OP.

Maybe that's an exercise in futility. Maybe we should expand the scope of this PR to fix more variants of the bug, and wrap them all in an `apiVersion` check?

File: api/src/opentrons/protocols/advanced_control/transfers.py (hunk 5 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocols/advanced_control/transfers.py:610]: Yeah, I'm thinking about this too.

If you think just swapping the type of `ValueError` to something more specific would help, I can certainly do that.

-----

I'm also trying to figure out how to get more user-friendly errors for people using the outer protocol API, like:

    ValueError: Can't distribute with a disposal_volume of 400 µL when the tip can only hold 300 µL.
    
As opposed to either of these:

    ValueError: max_volume must be greater than 0. (Got 0.)
    
    UnplannableTransferError: max_volume must be greater than 0. (Got 0.)

The tension is:

* `_expand_for_volume_constraints()` is central, so if we validate its arguments here, that validation can cover a lot of ground.
* But on the other hand, it's too low-level for that validation to raise user-friendly diagnostics. It can't know *why* `max_volume` is 0, so error messages will necessarily be opaque and jargony.

File: api/tests/opentrons/protocols/advanced_control/test_transfers.py (hunk 3 above)
Comment (amitlissack) [reply api/tests/opentrons/protocols/advanced_control/test_transfers.py]: I think it would be nice to explain this behavior now rather than leave the tests marked as skip.

Comment (amitlissack) [reply api/tests/opentrons/protocols/advanced_control/test_transfers.py]: While it's good that you added these tests, the general testing strategy for this module is tricky. I would have preferred to see individual tests for all the steps. **I'm not suggesting you do that now for all the functions**. But perhaps just for `_expand_for_volume_constraints`?

Comment (SyntaxColoring) [reply api/tests/opentrons/protocols/advanced_control/test_transfers.py]: > I would have preferred to see individual tests for all the steps.

Meaning:

* One test function to make sure `distribute()` errors when asked to use a`p20_single_gen2` with 20 µL tips to distribute with a disposal volume of 20 µL.
* One test function to make sure `distribute()` errors when asked to use a`p20_single_gen2` with 10 µL tips to distribute with a disposal volume of 20 µL.
* etc.

Or do I misunderstand?

**Comment by SyntaxColoring** [comment]
Based on review comments and out-of-band conversations, this bug doesn't seem like a good thing to attempt to solve incrementally.

The backwards compatibility concerns that make this such a slog to work on are ticketed for research in #7477.

Meanwhile, I'll re-draft this PR, try to understand the variants of this bug that I had deferred investigating, and see if we can fix them all in one `apiLevel`-bumping chunk.

File: api/tests/opentrons/protocols/advanced_control/test_transfers.py (hunk 3 above)
Comment (amitlissack) [reply api/tests/opentrons/protocols/advanced_control/test_transfers.py]: Sorry for not responding sooner. 

I meant that i'd want to see tests for the individual functions (`_create_volume_gradient` , `_check_valid_well_list`, `_create_volume_list`, `_expand_for_volume_constraints`  etc. ) rather than just full transfer plan.

**Comment by amitlissack** [comment]
> Based on review comments and out-of-band conversations, this bug doesn't seem like a good thing to attempt to solve incrementally.
> 
> The backwards compatibility concerns that make this such a slog to work on are ticketed for research in #7477.
> 
> Meanwhile, I'll re-draft this PR, try to understand the variants of this bug that I had deferred investigating, and see if we can fix them all in one `apiLevel`-bumping chunk.

I think that it would be good enough to dive potential other bug(s) and create tickets. The most important thing is to fix the infinite loop; either by raising an exception or doing nothing.

**Comment by SyntaxColoring** [comment]
I'm dropping this, for now. A combination of things makes this bug deceptively difficult to solve ~~correctly~~ safely.

This particular area of code:

* Has an enormous parameter surface area. (Think of all the possible ways of combining `disposal_volume`, `air_gap`, `distribute` vs. `consolidate` vs. `transfer`, Single-Channel vs. 8-Channel pipettes, transfer volume, pipette volume, tip volume, and [`apiLevel` backwards compatibility](https://github.com/Opentrons/opentrons/issues/7516).)
* Has never had parameter validation.
* Generally only has test coverage of its happy paths.

So, there's currently a big space of uncharacterized, undefined behavior that you can accidentally fall into. #6170, the bug that kicked this off, originally reported an infinite loop—but depending on the exact test case, instead of that, you can get other confusing results, like a big transfer breaking down into a `pick_up_tip` with no liquid handling steps.

Fully characterizing this space of undefined behavior turns out to be a lot of work. The work so far in this PR only partially accomplishes it.

But, by my understanding of our [Python Protocol API's versioning policy](https://docs.opentrons.com/v2/versioning.html), we *must* characterize and preserve that behavior in fixing these bugs. Even though it was undocumented, and didn't make sense, and only happened in incorrectly written protocols.

In short, I can't be confident that, if I fix these bugs, I won't unintentionally change behavior in some place where I'm not specifically looking.

-----

The more general versioning problem is ticketed for research as #7477. Depending on the outcome of that research, this could get a lot easier, and we could revisit it.

If/when we come back to this, I left `#todo` and `#fixme` comments in this branch to record the next steps.

And, anyway, we think we'll rewrite this code in Protocol Engine (the ongoing overhaul of our protocol-running back-end).
